api/src/main/java/org/apache/iceberg/Schema.java

```diff
@@ -314,4 +315,31 @@ public String toString() {
             .map(f -> "  " + f)
             .collect(Collectors.toList())));
   }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof Schema)) {
+      return false;
+    }
+
+    Schema that = (Schema) o;
+    if (!Objects.equals(struct.fields(), that.struct.fields())) {
+      return false;
+    }
+
+    if (aliasToId == that.aliasToId) {
+      return true;
+    }
+
+    if (aliasToId == null || that.aliasToId == null) {
+      return false;
+    }
+
+    return aliasToId.equals(that.aliasToId);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(struct.fields(), aliasToId == null ? 0 : aliasToId.hashCode());
+  }
```

Comment:
`Schema` doesn't implement `equals` because it isn't clear what schema equality means. Because we track fields by ID, two schemas are equal only if their fields have the same IDs, but most people don't think about schema equality that way and think of a SQL schema, like `id bigint, data string`. To avoid confusion, we don't provide an `equals` method that may have confusing results. Instead, we implement `equals` for structs so you can use `schema.asStruct()` in test assertions.